src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs

```diff
@@ -689,6 +689,114 @@ protected override void ProcessRecord()
         }
 
         #endregion Overrides
+
+        /// <summary>
+        /// Adds UserName, Description, BinPath, DelayedAutoStart and StartupType to a ServiceController object.
+        /// </summary>
+        /// <param name="service"></param>
+        /// <returns>ServiceController as PSObject with UserName, Description and StartupType added</returns>
+        private PSObject AddProperties(ServiceController service) {
+            NakedWin32Handle hScManager = IntPtr.Zero;
+            NakedWin32Handle hService = IntPtr.Zero;
+            int lastError = 0;
+            PSObject serviceAsPSObj = PSObject.AsPSObject(service);
+            IntPtr descriptionStructPtr = IntPtr.Zero;
+            IntPtr delayedAutoStartStuctPtr = IntPtr.Zero;
+            IntPtr serviceConfigStructPtr = IntPtr.Zero;
+            try
+            {
+                hScManager = NativeMethods.OpenSCManagerW(
+                    lpMachineName: service.MachineName,
+                    lpDatabaseName: null,
+                    dwDesiredAccess: NativeMethods.SC_MANAGER_CONNECT
+                );
+                if (IntPtr.Zero == hScManager) {
+                    lastError = Marshal.GetLastWin32Error();
+                    Win32Exception exception = new Win32Exception(lastError);
+                    WriteNonTerminatingError(
+                        service,
+                        service.MachineName,
+                        exception,
+                        "ComputerAccessDenied",
+                        ServiceResources.ComputerAccessDenied,
+                        ErrorCategory.PermissionDenied);
+                }
+                hService = NativeMethods.OpenServiceW(
+                    hScManager,
+                    service.ServiceName,
+                    NativeMethods.SERVICE_QUERY_CONFIG
+                );
+                if (IntPtr.Zero == hService) {
+                    lastError = Marshal.GetLastWin32Error();
+                    Win32Exception exception = new Win32Exception(lastError);
+                    WriteNonTerminatingError(
+                        service,
+                        exception,
+                        "CouldNotGetServiceInfo",
+                        ServiceResources.CouldNotGetServiceInfo,
+                        ErrorCategory.PermissionDenied);
+                }
+
+                bool queryStatus = NativeMethods.QueryServiceConfig2(hService, NativeMethods.SERVICE_CONFIG_DESCRIPTION, out descriptionStructPtr);
+                NativeMethods.SERVICE_DESCRIPTIONW description = (NativeMethods.SERVICE_DESCRIPTIONW)Marshal.PtrToStructure(descriptionStructPtr, typeof(NativeMethods.SERVICE_DESCRIPTIONW));
```

Comment:
I'd like to see these method calls organized differently to protect against potential memory leaks. The pattern is repeated in NativeMethods.QueryServiceConfig2 and I'd like all instances of it fixed. Since NativeMethods.QueryServiceConfig() is a helper function, its output parameter, instead of being an IntPtr structurePointer, should be the managed object NativeMethods.SERVICE_DESCRIPTIONW. That way, the unmanaged memory that is allocated within that function is also cleaned up within that function. The managed object carries the values out of the function and eventually gets garbage collected. To fix this instance, line 741 should be moved within the helper function. You can probably make Nativemethods.QueryServiceConfig2() a generic method as well to avoid having to use a switch for infoLevel.

Reply:
This sounds like a great idea and I'm all for it.
Could you describe a bit more what you would change? 
Like somewhere with an example.
The problem I'm having now is if I return the `struct` as an `out`, I have to set the variable before returning. I could set default values, but that means I have to check the values when adding the properties to my `Service`-object.

Reply:
@mirichmo, could you comment on my last comment?

Reply:
The idea is to put (NativeMethods.QUERY_SERVICE_CONFIG)Marshal.PtrToStructure(serviceConfigStructPtr, typeof(NativeMethods.QUERY_SERVICE_CONFIG)); and Marshal.FreeCoTaskMem in QueryServiceConfig2 and return the managed object NativeMethods.SERVICE_DESCRIPTIONW

Reply:
I had something like this in mind. The code is a bit messy and should be cleaned up, but illustrates the idea. The idea is that the helper function is a generic method that handles all the types in a similar manner. It also keeps allocation and freeing within the same method.

### Method

```C#
internal static bool QueryServiceConfig2<T>(NakedWin32Handle hService, DWORD infolevel, out T configStructure)
{
            IntPtr lpBuffer = IntPtr.Zero;
            structurePointer = IntPtr.Zero;
            DWORD bufferSize, bufferSizeNeeded = 0;

            bool status = NativeMethods.QueryServiceConfig2W(
                hService: hService,
                dwInfoLevel: infolevel,
                lpBuffer: lpBuffer,
                cbBufSize: 0,
                pcbBytesNeeded: out bufferSizeNeeded);

            if (status != true && Marshal.GetLastWin32Error() != ERROR_INSUFFICIENT_BUFFER) {
                return status;
            }

    try {
            lpBuffer = Marshal.AllocCoTaskMem((int)bufferSizeNeeded);
            bufferSize = bufferSizeNeeded;
           status = NativeMethods.QueryServiceConfig2W(
               hService,
               infolevel,
                lpBuffer,
                bufferSize,
                out bufferSizeNeeded);

        configStructure = (T)Marshal.PtrToStructure(lpBuffer, typeof(T));
    }
    finally
    {
        Marshal.FreeCoTaskMem(lpBuffer);
    }
            return status;
}
```

### Caller

```C#
NativeMethods.SERVICE_DESCRIPTIONW descriptionStruct = new NativeMethods.SERVICE_DESCRIPTIONW();

bool querySuccessful = NativeMethods.QueryServiceConfig2<NativeMethods.SERVICE_DESCRIPTIONW>(hService, NativeMethods.SERVICE_CONFIG_DESCRIPTION, out descriptionStruct);

// No free is needed here
```

Reply:
Thank you! 👍  I made the updates. Could you please review @mirichmo